torch/distributed/_tools/fsdp2_memory_tracker.py

```diff
@@ -0,0 +1,539 @@
+            divisor = 2**10
+        return divisor
+
+    def _rounding_fn(self, value: int, precision: int) -> Union[float, int]:
+        divisor = self._get_mem_divisor()
+        if divisor == 1:
+            return value
+        return round(value / divisor, precision)
+
+    def print_mem_stats(self, stats: Optional[Dict[str, int]] = None):
+        if stats is None:
+            stats = self._get_current_memory_allocated()
+
+        for mem_type, mem_val in stats.items():
+            print(
+                f"\t{mem_type}:"
+                f" {self._rounding_fn(mem_val, 2)} {self.units}"
+            )
+
+    def get_max_memory(self) -> int:
+        return self._MAX_MEMORY
+
+    def _display_peak_mem_stats(self):
+        print(
+            f"Peak Memory Usage: {self._rounding_fn(self._MAX_MEMORY, 2)} {self.units}"
+        )
+        print("Peak Memory Usage Breakdown:")
+        self.print_mem_stats(self._MAX_MEMORY_BREAKDOWN)
+
+    def _display_modulewise_mem_stats(self):
+        for mod in self.memory_tracking.keys():
+            print(f"Module:  {mod}")
+            for state, stats in self.memory_tracking[mod].items():
+                print(f"{state}")
+                self.print_mem_stats(stats)
+            print()
+
+    def _enter_module(self, name: str, state: str):
+        self.parents.append(name)
+        self.memory_tracking[name][
+            state
+        ] = self._get_current_memory_allocated()
+
+    def _exit_module(self, name: str, state: str):
+        assert self.parents[-1] == name, f"{self.parents[-1]} is not {name}"
+        self.memory_tracking[name][
+            state
+        ] = self._get_current_memory_allocated()
+        self.parents.pop()
+
+    def _instrument_fsdp_param_group(self, fsdp_param_group: FSDPParamGroup):
+
+        for fsdp_param in fsdp_param_group.fsdp_params:
+            assert isinstance(fsdp_param.sharded_param, nn.Parameter), (
+                f"{fsdp_param._module_info.param_name} sharded param is not a"
+                " nn.Parameter"
+            )
+            self._update_and_maybe_create_winfo(
+                fsdp_param.sharded_param, _RefType.sharded_parameter
+            )
+
+    def record_pre_forward_order(self, fsdp_param_group: FSDPParamGroup):
+        self.pre_forward_order.append(fsdp_param_group)
+        self.pre_forward_order_indices.append(len(self.pre_forward_order) - 1)
+
+    def _fsdp_param_group_pre_forward(
+        self,
+        orig_fsdp_param_group_pre_forward: Callable,
+        fsdp_param_group: FSDPParamGroup,
+        name: str,
+    ) -> Callable:
+        @wraps(orig_fsdp_param_group_pre_forward)
+        def inner(*args, **kwargs):
+            self._enter_module(name, "Before Pre-Forward")
+            args, kwargs = orig_fsdp_param_group_pre_forward(*args, **kwargs)
+            for fsdp_param in fsdp_param_group.fsdp_params:
+                assert isinstance(fsdp_param._unsharded_param, nn.Parameter), (
```

Comment:
Only nit from our discussion was to try to use the public `fsdp_param.unsharded_param` here instead of `_unsharded_param` since it will include a check that the parameter is in the unsharded state.